### ISSUE TITLE
argparse: Add the keyword argument `subspace_dest` of type `Optional[str]` to `argparse._SubParsersAction.add_parser` to allow sub-namespaces which correspond to subparsers

### ISSUE THREAD

**Issue by jb2170** [python/cpython#103639]
# Enhancement

This enhancement is completely forwards/backwards compatible.

We add the following: Check for the presence of `subspace_dest` in the `kwargs` of `argparse._SubParsersAction.add_parser(self, name, **kwargs)`. If present we store it as the attribute `subspace_dest` on the subparser created. When a subparser `sp` is created via `argparse._SubParsersAction.add_parser(..., subspace_dest='foo')`, arguments parsed by this subparser are be stored in another `Namespace` object, which is stored within the parent parser's `Namespace` object as the attribute with name `'foo'`.

# Pitch

This is a beneficial enhancement for many reasons:

- Avoid shadowing / overwriting of argument destinations within a `Namespace`. Previously, if arguments in a subparser wanted to be sure they were not colliding with argument names from parent parsers, their `dest` parameters would have to be prefixed appropriately. Now arguments can be added to an argument parser agnostically, without care for prefixing `dest` parameters; they are relative to the parser's `Namespace`, not a 'global' parent `Namespace`. This also simplifies the `dest` parameters.
- One may have a program which splits into subprograms, each one of which only needs to know about arguments corresponding to its subparser. Each subprogram may define a set of arguments stored in its own ArgumentParser, which the overarching master program may combine into one argument parser via `ArgumentParser.add_subparsers`. Any `dest` parameters specified in each of the subprogram's parsers can be done without fear of colliding with sibling parsers. After parsing the CLI arguments, the sub-`Namespace` can be passed to the corresponding subprogram, as will be demonstrated below.

```py
#!/usr/bin/env python3

import argparse

def get_cli_args(argv) -> argparse.Namespace:
    parser = argparse.ArgumentParser(prog="doorctl", description = "open and close doors")
    parser_action = parser.add_subparsers(title="action", dest="action", required=True)

    parser_action_open = parser_action.add_parser("open")
    parser_action_open.add_argument("action_open_name", metavar="NAME")

    parser_action_close = parser_action.add_parser("close")
    parser_action_close.add_argument("action_close_name", metavar="NAME")

    return parser.parse_args(argv)

def get_cli_args_subspace(argv) -> argparse.Namespace:
    parser = argparse.ArgumentParser(prog="doorctl", description = "open and close doors")
    parser_action = parser.add_subparsers(title="action", dest="action", required=True)

    parser_action_open = parser_action.add_parser("open", subspace_name="action_open")
    parser_action_open.add_argument("name", metavar="NAME")

    parser_action_close = parser_action.add_parser("close", subspace_name="action_close")
    parser_action_close.add_argument("name", metavar="NAME")

    return parser.parse_args(argv)

def main() -> None:
    argv = ["open", "front"]

    print(get_cli_args(argv))
    # Namespace(action='open', action_open_name='front')

    print(get_cli_args_subspace(argv))
    # Namespace(action='open', action_open=Namespace(name='front'))

if __name__ == "__main__":
    main()
```
Another example is given below

```py
#!/usr/bin/env python3

import argparse

def add_cli_args_common_af(parser: argparse.ArgumentParser) -> None:
    # common between `echo` and `write`
    parser_af = parser.add_subparsers(title="address", dest="address_family", required=True,
        description="which address family (AF) to create the socket in"
    )

    parser_af_unix = parser_af.add_parser("unix", help="AF_UNIX", subspace_name="unix")
    parser_af_unix.add_argument("path", metavar="SOCKET-PATH")

    parser_af_inet = parser_af.add_parser("inet", help="AF_INET", subspace_name="inet")
    parser_af_inet.add_argument("ip", metavar="IP")
    parser_af_inet.add_argument("port", metavar="PORT", type=int)

def add_cli_args_echo(parser: argparse.ArgumentParser) -> None:
    add_cli_args_common_af(parser)

def add_cli_args_write(parser: argparse.ArgumentParser) -> None:
    add_cli_args_common_af(parser)

def add_cli_args(parser: argparse.ArgumentParser) -> None:
    parser_action = parser.add_subparsers(title="action", dest="action", required=True)

    add_cli_args_echo(parser_action.add_parser("echo",
        help="Connect to a socket and send recv-ed data back to it",
        subspace_name="echo"
    ))

    add_cli_args_write(parser_action.add_parser("write",
        help="Connect to a socket and read+write data with it. \
              sends stdin to the socket and recvs the socket to stdout",
        subspace_name="write"
    ))

def get_cli_args(argv) -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="socket-tool")

    add_cli_args(parser)

    args = parser.parse_args(argv)

    return args

def use_cli_args_echo(args: argparse.Namespace):
    # Pass args to an implementation of an echo server
    print(f"/echo:      {args}")
    if args.address_family == "inet":
        print(f"/echo/inet: {args.inet}")
    else:
        print(f"/echo/unix: {args.unix}")

def use_cli_args_write(args: argparse.Namespace):
    # Pass args to an implementation of a polling of stdin and a socket
    print(f"/write:      {args}")
    if args.address_family == "inet":
        print(f"/write/inet: {args.inet}")
    else:
        print(f"/write/unix: {args.unix}")

def use_cli_args(args: argparse.Namespace):
    if args.action == "echo":
        use_cli_args_echo(args.echo)
    elif args.action == "write":
        use_cli_args_write(args.write)

def main() -> None:
    argv = ["write", "inet", "127.0.0.1", "80"]

    args = get_cli_args(argv)

    print(f"/:           {args}")

    use_cli_args(args)

if __name__ == "__main__":
    main()
```

This outputs

```
/:           Namespace(action='write', write=Namespace(address_family='inet', inet=Namespace(ip='127.0.0.1', port=80)))
/write:      Namespace(address_family='inet', inet=Namespace(ip='127.0.0.1', port=80))
/write/inet: Namespace(ip='127.0.0.1', port=80)
```

Notice the modularity of the code, and simple `dest` parameters for both positional and keyword arguments, both in adding arguments to parsers, and getting arguments from `Namespace`s

## Relation to other requests

This supersedes Issue #59633 aka BPO 15428, a solution to which suggests using declinable `dest` parameters eg `dest='test.doo.deep'`. Our solution is much more modular and flexible, as our `dest` parameters / subnamespaces are relative, whereas BPO 15428 suggests absolute subnamespaces.

Documentation is included in my following PR corresponding to this issue

### Linked PRs
* gh-103640

**Comment by hpaulj** [comment]
It might be a good idea to implement this as a new subclass of  `_SubParsersAction`.  Then at least during development, both the default Action class will be available as well as this new one.  I need to read your proposed changes but I think they all occur in the `_SubParsersAction` class, mainly its `__call__` and `add_parser` methods.

`add_subparsers` uses the `registry` to find the correct Action class.  So on a temporary basis the registry could be changed to activate the changes.

**Comment by jb2170** [comment]
> ...current version... This gives the subparser defaults priority

Indeed. Here is an example of using subparsers which includes conflicting `dest` parameters in the current behaviour, but not with the proposed changes

```py
import argparse

def get_parser(subspace_name):
    parser = argparse.ArgumentParser()
    parser_what = parser.add_subparsers(title="what", dest="what", help="what do to", required=True)

    parser_what_cat = parser_what.add_parser("cat", subspace_name=subspace_name)
    parser_what_cat.add_argument("what", help="what to cat")

    # more parser_what_X could be added here. omitted for brevity of the example
    return parser

print(get_parser(None).parse_args(["cat", "/dev/null"]))
# Namespace(what='/dev/null')

print(get_parser("cat").parse_args(["cat", "/dev/null"]))
# Namespace(what='cat', cat=Namespace(what='/dev/null'))
```

Using nested `Namespace`s is avoids collisions, as raised in other issues.

> My general advise has been to avoid using the same dest in both the main and subparsers

Ah the days of old DOS with its flat filesystem. Unix was miles ahead with its hierarchical layout. To avoid conflicting `dest` parameters I have always used a meticulous system of `subparserArg_choiceX_subparserArg_choiceX...` of which an example can be seen below. However this leads to extremely long prefixed `dest` values, and requires one to keep track of n-layers of nested subparsers. Indeed in `example_with` below observe that with nested `Namespace`s the `dest` parameter is sometimes not even needed. This is because in the `example_with` function the `dest` parameters for the flags are 'local' with respect to the corresponding `Namespace`.

```py
import ap310 as argparse

def example_without(argv):
    parser = argparse.ArgumentParser()

    parser_decision = parser.add_subparsers(title="decision", dest="decision", required=True)

    parser_decision_add = parser_decision.add_parser("add")
    parser_decision_add_decision = parser_decision_add.add_subparsers(title="decision", dest="decision_add_decision", required=True)

    parser_decision_add_decision_eat = parser_decision_add_decision.add_parser("eat")
    parser_decision_add_decision_eat.add_argument("--vegetarian", action="store_true", dest="decision_add_decision_eat_vegetarian")

    parser_decision_add_decision_sleep = parser_decision_add_decision.add_parser("sleep")
    parser_decision_add_decision_sleep.add_argument("--snore", action="store_true", dest="decision_add_decision_sleep_snore")

    parser_decision_remove = parser_decision.add_parser("remove")
    parser_decision_remove_decision = parser_decision_remove.add_subparsers(title="decision", dest="decision_remove_decision", required=True)

    parser_decision_remove_decision_leave = parser_decision_remove_decision.add_parser("leave")
    parser_decision_remove_decision_leave.add_argument("--discreetly", action="store_true", dest="decision_remove_decision_leave_discreetly")

    parser_decision_remove_decision_read = parser_decision_remove_decision.add_parser("read")
    parser_decision_remove_decision_read.add_argument("--aloud", action="store_true", dest="decision_remove_decision_read_aloud")

    return parser.parse_args(argv)

    """
    decision
    decision_add_decision
    decision_add_decision_eat_vegetarian
    decision_add_decision_sleep_snore
    decision_remove_decision
    decision_remove_decision_leave_discreetly
    decision_remove_decision_read_aloud
    """

def example_with(argv):
    parser = argparse.ArgumentParser()

    parser_decision = parser.add_subparsers(title="decision", dest="decision", required=True)

    parser_decision_add = parser_decision.add_parser("add", subspace_name="add")
    parser_decision_add_decision = parser_decision_add.add_subparsers(title="decision", dest="decision", required=True)

    parser_decision_add_decision_eat = parser_decision_add_decision.add_parser("eat", subspace_name="eat")
    parser_decision_add_decision_eat.add_argument("--vegetarian", action="store_true")

    parser_decision_add_decision_sleep = parser_decision_add_decision.add_parser("sleep", subspace_name="sleep")
    parser_decision_add_decision_sleep.add_argument("--snore", action="store_true")

    parser_decision_remove = parser_decision.add_parser("remove", subspace_name="remove")
    parser_decision_remove_decision = parser_decision_remove.add_subparsers(title="decision", dest="decision", required=True)

    parser_decision_remove_decision_leave = parser_decision_remove_decision.add_parser("leave", subspace_name="leave")
    parser_decision_remove_decision_leave.add_argument("--discreetly", action="store_true")

    parser_decision_remove_decision_read = parser_decision_remove_decision.add_parser("read", subspace_name="read")
    parser_decision_remove_decision_read.add_argument("--aloud", action="store_true")

    return parser.parse_args(argv)

    """
    decision               # subparsers       # attr
    `-add                  #   parser         # Namespace
    | `-decision           #     subparsers   #   `-attr
    |   `-eat              #       parser     #   `-Namespace
    |   | `-vegetarian     #         option   #   | `-attr
    |   `-sleep            #       parser     #   `-Namespace
    |     `-snore          #         option   #     `-attr
    `-remove               #   parser         # Namespace
      `-decision           #     subparsers   #   `-attr
        `-leave            #       parser     #   `-Namespace
        | `-discreetly     #         option   #   | `-attr
        `-read             #       parser     #   `-Namespace
          `-aloud          #         option   #     `-attr
    """

argv = ["add", "sleep", "--snore"]

print(example_without(argv))
# Namespace(decision='add', decision_add_decision='sleep', decision_add_decision_sleep_snore=True)

print(example_with(argv))
# Namespace(decision='add', add=Namespace(decision='sleep', sleep=Namespace(snore=True)))
```

Introducing an (optional) hierarchy is a very good quality-of-life addition 

> a new subclass of _SubParsersAction

This isn't necessary as the new keyword `subspace_name` argument is optional / n-wards compatible, and all that occurs when `subspace_name` is specified is the `subnamespace` is set as an attribute of the parent `namespace`, instead of copying / flattening / overwriting `subnamespace`'s attributes into `namespace`

> I need to read your proposed changes

Please do, there's not much to it

**Comment by jb2170** [comment]
TLDR-ing my previous comment, observe how much neater a hierarchical-filesystem-like layout is, without the need for absolute prefixes to ensure no `dest`-collisions

Before
```
decision
decision_add_decision
decision_add_decision_eat_vegetarian
decision_add_decision_sleep_snore
decision_remove_decision
decision_remove_decision_leave_discreetly
decision_remove_decision_read_aloud
```

After
```
dest                     parser type        Namespace or attr

decision               # subparsers       # attr
`-add                  #   parser         # Namespace
| `-decision           #     subparsers   #   `-attr
|   `-eat              #       parser     #   `-Namespace
|   | `-vegetarian     #         option   #   | `-attr
|   `-sleep            #       parser     #   `-Namespace
|     `-snore          #         option   #     `-attr
`-remove               #   parser         # Namespace
  `-decision           #     subparsers   #   `-attr
    `-leave            #       parser     #   `-Namespace
    | `-discreetly     #         option   #   | `-attr
    `-read             #       parser     #   `-Namespace
      `-aloud          #         option   #     `-attr
```

**Comment by jb2170** [comment]
Stale. Would discuss on https://discuss.python.org if interest peaks again.

**Comment by jb2170** [comment]
Discussion thread

https://discuss.python.org/t/107202
